### PR TITLE
feat(deploy): add github pages deploy

### PR DESCRIPTION
**Pull request by filipesilva**
This PR adds (and improves upon) the functionality previously in https://github.com/IgorMinar/angular-cli-github-pages directly to the CLI. 

Later, this functionality will be split into an addon of it's own again, but for now it is better to keep it in the same codebase for rapid development.

### CODE REVIEW

File: addon/ng2/commands/github-pages-deploy.ts [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+import * as Promise from 'ember-cli/lib/ext/promise';
+import * as chalk from 'chalk';
+import * as fs from 'fs';
+import * as fse from 'fs-extra';
+import * as path from 'path';
+import * as BuildTask from 'ember-cli/lib/tasks/build';
+import * as win from 'ember-cli/lib/utilities/windows-admin';
+import * as CreateGithubRepo from '../tasks/create-github-repo'
+
+// import * as https from 'https';
```
Comment (hansl) [inline addon/ng2/commands/github-pages-deploy.ts]: Please remove commented code.

File: addon/ng2/commands/github-pages-deploy.ts [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+
+  run: function(options, rawArgs) {
+    var ui = this.ui;
+    var root = this.project.root;
+    var execOptions = {
+      cwd: root
+    };
+    var projectName = this.project.pkg.name;
+
+    var fsReadFile = Promise.denodeify(fs.readFile);
```
Comment (hansl) [inline addon/ng2/commands/github-pages-deploy.ts]: These should be in the global scope, and `const`.

File: addon/ng2/commands/github-pages-deploy.ts [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+    }
+
+    function pushToGitRepo(committed) {
+      return execPromise(`git push origin ${options.branch}`);
+    }
+
+    function printProjectUrl() {
+      return execPromise('git remote -v')
+        .then((stdout) => {
+        let userName = stdout.match(/origin\s+git@github\.com\:([^\/]+)/m)[1].toLowerCase();
```
Comment (hansl) [inline addon/ng2/commands/github-pages-deploy.ts]: Did you lint?  Indentation is wrong here.

File: addon/ng2/index.js [hunk 4]
```diff
@@ -1,4 +1,4 @@
-/* jshint node: true */
+ /* jshint node: true */
```
Comment (hansl) [inline addon/ng2/index.js]: Remove space please.

File: addon/ng2/tasks/create-github-repo.ts [hunk 5]
```diff
@@ -0,0 +1,66 @@
+import * as Promise from 'ember-cli/lib/ext/promise';
+import * as Task from 'ember-cli/lib/models/task';
+import { exec } from 'child_process';
+import * as https from 'https';
+import * as inquire from 'inquirer';
+
+module.exports = Task.extend({
+  run: function(commandOptions) {
+    var ui = this.ui;
+    var execPromise = Promise.denodeify(exec);
```
Comment (hansl) [inline addon/ng2/tasks/create-github-repo.ts]: Again, this should be at the global scope, and `const`.

**Comment by hansl** [comment]
For testing, verify if you can monkey patch exec like this: `require('child_process').exec = function(x, callback) { /*...*/; callback(); }`. That should make it possible to bypass everything in the shell.

File: package.json [hunk 6]
```diff
@@ -40,6 +40,7 @@
     "ember-cli": "2.4.2",
     "exit": "^0.1.2",
     "fs-extra": "^0.26.6",
+    "inquirer": "^0.10.1",
```
Comment (jkuri) [inline package.json]: `inquirer` comes with `ember-cli`, we don't need another package.

Comment (jkuri) [inline package.json]: FYI you can use it as `return this.ui.prompt({ ...`.

File: addon/ng2/commands/github-pages-deploy.ts (hunk 1 above)
Comment (filipesilva) [inline addon/ng2/commands/github-pages-deploy.ts]: Ugh.. my bad. Done.

File: addon/ng2/commands/github-pages-deploy.ts (hunk 2 above)
Comment (filipesilva) [inline addon/ng2/commands/github-pages-deploy.ts]: Done.

File: addon/ng2/index.js (hunk 4 above)
Comment (filipesilva) [inline addon/ng2/index.js]: Done.

File: addon/ng2/commands/github-pages-deploy.ts (hunk 3 above)
Comment (filipesilva) [inline addon/ng2/commands/github-pages-deploy.ts]: My beautifier was acting a bit strange for some reason, not sure why. Fixed.

File: addon/ng2/tasks/create-github-repo.ts (hunk 5 above)
Comment (filipesilva) [inline addon/ng2/tasks/create-github-repo.ts]: Done.

File: package.json (hunk 6 above)
Comment (filipesilva) [inline package.json]: Did not know that! Using it now.

**Comment by filipesilva** [comment]
Fixed the things pointed out in the comments, looking at tests now.

**Comment by hansl** [comment]
@filipesilva: ping. What's the status about testing this?

**Comment by hansl** [comment]
@filipesilva: ping. What's the status about testing this?

**Comment by filipesilva** [comment]
@hansl been working on tests yesterday and today. Expect to have a good few tests by the end of tomorrow.

Was a bit tricky since I had to stub `exec` but also provide it with a list of command/response pairs (including error'd ones) for each test. This is because the results determine the path the command/task code will take. 

I plan to put this out as a helper for other tests to use.

**Comment by filipesilva** [comment]
@hansl tests are in.

File: README.md [hunk 7]
```diff
@@ -201,15 +210,6 @@ This will use the `lint`/`format` npm script that in generated projects uses `ts
 
 You can modify the these scripts in `package.json` to run whatever tool you prefer.
 
-
-### Formatting code
```
Comment (filipesilva) [inline README.md]: Kinda piggybacking on this PR tbh... but I noticed this is outdated as the previous section covers it. Can put it on a different PR if needed.

**Comment by hansl** [comment]
LGTM.
